### PR TITLE
Sync writes __source__ falsely into frozen layers

### PR DESCRIPTION
**Pull request by Wiseqube**
This PR fixes an issue where zodbsync was overwriting the contents of a source-file in a frozen layer. 

The test-case is as follows:
1. We record an object into the custom layer
2. The filesystem objects are moved to the frozen layer manually
3. The content in the Data.FS is modified, in this case we edith both the meta aswell as the content data
4. A record is executed. 

The expected behaviour would be that the changed meta and source files are both written into the custom-layer. The actual behaviour is that the `__meta__`-File is correctly written into the custom-layer but the `__source__`-File is modified in the frozen layer below!

The issue was that the path which was generated for the source-file was based on the layer-path while the path for the meta-file was generated based upon the custom upper layer!

### CODE REVIEW

**Review by gandie — COMMENTED** [review]
One typo, also i suggest being extra suspicious:

```diff
diff --git a/perfact/zodbsync/tests/test_sync.py b/perfact/zodbsync/tests/test_sync.py
index 0c27784..f545d6c 100644
--- a/perfact/zodbsync/tests/test_sync.py
+++ b/perfact/zodbsync/tests/test_sync.py
@@ -2179,6 +2179,10 @@ class TestSync():
             # both meta and source file are in custom layer
             assert os.path.exists(os.path.join(root, 'blob/__meta__'))
             assert os.path.exists(os.path.join(root, 'blob/__source__.txt'))
-            with open('{}/__root__/blob/__source__.txt'.format(layer)) as f:
+            source_fmt = '{}/__root__/blob/__source__.txt'
+            with open(source_fmt.format(layer)) as f:
                 # source in layer should still be empty
                 assert f.read() == ''
+            with open(source_fmt.format(self.repo.path)) as f:
+                # ... content is in custom layer!
+                assert f.read() == 'text_content'
```

File: perfact/zodbsync/tests/test_sync.py [hunk 1]
```diff
@@ -2152,3 +2153,32 @@ def test_layer_update_warn(self, caplog):
             assert expect + '/Test' in caplog.text
             assert 'AttributeError' not in caplog.text
             assert expect + '/ToDelete/Sub' in caplog.text
+
+    def test_layer_frozen(self):
+        """
+        Verify that changed files are propery written into the custom
```
[the remark this answers is not in the document]
Comment (Wiseqube) [reply perfact/zodbsync/tests/test_sync.py]: Thanks for the input. I've implemented both changes!

**Review by gandie — COMMENTED** [review]
What about this one to make things a little more clear (tests seem cool with it):

```diff
diff --git a/perfact/zodbsync/zodbsync.py b/perfact/zodbsync/zodbsync.py
index c0eaf01..85cf1c0 100644
--- a/perfact/zodbsync/zodbsync.py
+++ b/perfact/zodbsync/zodbsync.py
@@ -476,7 +476,7 @@ class ZODBSync:
         if old_data != new_data:
             # Path in top layer, might be different than the one where we read
             # the content
-            write_base = os.path.join(self.app_dir, path.lstrip('/'))
+            write_base = self.fs_path(path)
             os.makedirs(write_base, exist_ok=True)

             self.logger.debug("Will write %d bytes of metadata" % len(fmt))
```

Comment and docstring of `fs_path` then match nicely imo

**Comment by Wiseqube** [comment]
@gandie accidentially put this additional reafactoring into a branch which is already merged now, so I guess ready for review again!

File: perfact/zodbsync/tests/test_sync.py [hunk 2]
```diff
@@ -2152,3 +2153,36 @@ def test_layer_update_warn(self, caplog):
...
+
+    def test_layer_frozen(self):
+        """
+        Verify that changed files are properly written into the custom
+        layer in case the layer below is frozen.
+        """
+        with self.runner.sync.tm:
+            self.app.manage_addProduct['OFSP'].manage_addFile(id='blob')
+
+        with self.addlayer(frozen=True) as layer:
```
Comment (viktordick) [inline perfact/zodbsync/tests/test_sync.py:2165]: Pretty sure that `frozen = True` is the default for layers below the top-most custom layer. So not sure why this was necessary, but it still is OK.

**Review by viktordick — APPROVED** [review]
LGTM

**Pull request merged**